### DISCUSSION TITLE
What are the node energies?

### DISCUSSION THREAD

**Discussion by binxi0629** [ACEsuit/mace#417] [Q&A] [answered]
Hi, 

Good job for MACE-OFF23. I am using the implemented MACE ASE Calculator, but I am wondering what are the "node_energy" and `node_e0`? Are the node energies simply per-atom energy (but seems the sum of all node energies is not equal to the "energy")? What are the difference between `self.results['energy']` and `self.results["node_energy"]` ?  
```
self.results["node_energy"] = (
                torch.mean(ret_tensors["node_energy"] - node_e0, dim=0).cpu().numpy()
            )
```

**Comment by binxi0629** [comment]
BTW, if I modified it by adding `ret_tensors["node_energy"]` and `node_e0`, then the sum among all node energies (`self.results["node_energy"]`) is exactly the same as the total energy (`self.results['energy']`).

**Comment by ilyes319** [comment]
Hey! MACE predicts energies on each atoms that we call site energy, and the total energy of the system is the sum of these site energies. The `self.results["node_energy"]` are just the site energies. The `node_e0` corresponds to the isolated atoms energies, meaning what energy will MACE output for atoms that is alone in a vacuum.

**Comment by binxi0629** [comment]
Thanks for answering. My question is that when I am using the `mace_off` to calculate total energy and node energies for a given molecule conformation,  the sum of the site energies that I obtained is not the same as the output of the total energy. 

Taking H2O as an example, the following code gives the results:
```
from mace.calculators import mace_off
from ase import build
import numpy as np

atoms = build.molecule('H2O')
calc = mace_off(model="medium", device='cuda:1') # small, medium, large
atoms.calc = calc
tot_e = atoms.get_potential_energy()
node_energies = atoms.calc.results['node_energy']

tot_e, np.sum(node_energies)

#My output: (-2081.11639716893, 2061.03884806397) 
```

**Reply by RokasEl** [reply, marked as answer]
Hi, sorry for the late reply.

There was a bug in the calculator class, when calculating the node energies see https://github.com/ACEsuit/mace/pull/426.

Nonetheless, there still is a difference between the `get_potential_energy()` and the node energies.

`atoms.get_potential_energy()` includes the E0s of the isolated atoms.

`node_energies = atoms.calc.results['node_energy']` is the node energy **not including the isolated atom energy**. 

In other words, sum of the node energies is the energy relative to isolated atoms, and `atoms.get_potential_energy()` is the "absolute energy".
